src/compiler/parser.ts

```diff
@@ -1811,7 +1811,7 @@ namespace ts {
         function parseEntityName(allowReservedWords: boolean, diagnosticMessage?: DiagnosticMessage): EntityName {
             let entity: EntityName = parseIdentifier(diagnosticMessage);
             while (parseOptional(SyntaxKind.DotToken)) {
-                const node = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);
+                const node: QualifiedName = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);  // !!!

```

Comment:
@JsonFreeman we can't be too conservative, though - for example, in my example above, if I omitted 
```ts
    if (isThird(x)) {
        break;
    }
```
from the while loop, then `next(x)` could return `undefined` (as `x` could be `Three` at the start of the loop) - which isn't assignable to `x` under strict null mode - which should cause an error within the loop! Making it non-cumbersome to use overloads like this could be difficult without a complete fix for the issue.